### ISSUE TITLE
** 3.9 Beta testing **

### ISSUE THREAD

**Issue by Dreamsorcerer** [fafhrd91/aiohttp#7675]
**3.9.0rc0 is now out, likely to be promoted as a stable release at the weekend.**

We've introduced a few significant changes in 3.9, so it would be great to get some community testing before promoting it as a stable release. We've released 3.9.0b1 as a prerelease to help with this.

Can be installed with: `pip install 'aiohttp<4' --pre --upgrade`

In particular, these are things we are interested on testing or gathering feedback on:

**Client-side**

- #7764 - This fix resulted in changing a few things internally, so looking to see if there might be any issues around connection cleanup.

**Server-side**

- `AppKey` for static typing support: https://docs.aiohttp.org/en/latest/web_advanced.html#application-s-config
- Graceful shutdown on run_app(), which includes run_app() allowing some time for existing tasks (such as endpoint handlers) to complete before proceeding with the shutdown process: https://docs.aiohttp.org/en/latest/web_advanced.html#graceful-shutdown
- Handler cancellation option (this used to be the default behaviour a few releases ago): https://docs.aiohttp.org/en/latest/web_advanced.html#web-handler-cancellation

------

TODO: Prepare aiohttp libraries for AppKey:

- [x] aiohttp-admin (https://github.com/aio-libs/aiohttp-admin/pull/808)
- [x] aiohttp-debugtoolbar (https://github.com/aio-libs/aiohttp-debugtoolbar/pull/505)
- [ ] aiohttp-demos (can be done later)
- [x] aiohttp-devtools (https://github.com/aio-libs/aiohttp-devtools/pull/621)
- [x] aiohttp-jinja2 (https://github.com/aio-libs/aiohttp-jinja2/pull/534)
- [x] aiohttp-security (https://github.com/aio-libs/aiohttp-security/pull/671)
- [x] aiohttp-session (https://github.com/aio-libs/aiohttp-session/pull/943)
- [x] aiojobs (https://github.com/aio-libs/aiojobs/pull/460)

**Comment by Dreamsorcerer** [comment]
One thing that bothers me about 3.12 is that our import time test always fails in the initial test run, and then passes in the rerun. I'm not sure what change in 3.12 would cause that, but it seems to happen 100% of the time. It's not a small margin either, it misses the target time by around 30-50%.

Maybe someone else has some ideas?

**Comment by h-vetinari** [comment]
I spent some time to enable the test suite on our conda-forge "feedstock" that repackages aiohttp - unfortunately this wasn't turned on so far, most likely because it's a somewhat non-trivial effort, especially as aiohttp is really cumbersome to build 100% from source (I gave up on that after ~5 onion layers and used a work-around that's... kinda acceptable, but bespoke).

Compared to the "existing" failures (w.r.t. our infrastructure) for 3.8.6, this is the delta for [3.9.0b0](https://github.com/conda-forge/aiohttp-feedstock/pull/76):
```diff
@@ -60,14 +60,17 @@ test:
     {% set tests_to_skip = "_not_a_real_test" %}
-    {% set tests_to_skip = tests_to_skip + " or test_feed_eof_no_err_brotli" %}
     {% set tests_to_skip = tests_to_skip + " or test_http_response_parser_bad_chunked_strict_c[pyloop]" %}
     {% set tests_to_skip = tests_to_skip + " or test_http_response_parser_bad_chunked_strict_py[pyloop]" %}
     {% set tests_to_skip = tests_to_skip + " or test_http_response_parser_strict_headers[c-parser-pyloop]" %}
-    {% set tests_to_skip = tests_to_skip + " or test_request_tracing_url_params[pyloop]" %}
     {% set tests_to_skip = tests_to_skip + " or test_tcp_connector_raise_connector_ssl_error[pyloop]" %}
     # test failing with python >=3.11
     {% set tests_to_skip = tests_to_skip + " or test_https_proxy_unsupported_tls_in_tls[pyloop-https]" %}   # [py>=311]
+    # fails to open socket at desired IP/port
+    {% set tests_to_skip = tests_to_skip + " or (test_web_server and test_handler_cancellation)" %}         # [linux]
+    {% set tests_to_skip = tests_to_skip + " or (test_web_server and test_no_handler_cancellation)" %}      # [linux]
+    # emulation is too slow for expected speed
+    {% set tests_to_skip = tests_to_skip + " or (test_imports and test_import_time)" %}                     # [aarch64 or ppc64le]
     # aio seems to hard-code expectation for cpython (i.e. not "pypy3.9" in SP_DIR path)
     {% set tests_to_skip = tests_to_skip + " or test_static_route_user_home" %}             # [python_impl == "pypy"]
     # not working with pypy (all platforms)
     [...]
```

The timeout failure is unsurprising, as we have no native hardware for aach64/ppc64le, and so those run through emulation in QEMU, which is obviously slower.

The other two look like this:
```
FAILED sources/tests/test_web_server.py::test_handler_cancellation - OSError: [Errno 99] error while attempting to bind on address ('::1', 51455, 0, 0): cannot assign requested address
FAILED sources/tests/test_web_server.py::test_no_handler_cancellation - OSError: [Errno 99] error while attempting to bind on address ('::1', 56597, 0, 0): cannot assign requested address
```

All in all, I think this looks pretty OK 👍

**Comment by haukex** [comment]
I can see how the "It is recommended to use web.AppKey instances for keys" warning ([here](https://github.com/aio-libs/aiohttp/blob/3d60762a/aiohttp/web_app.py#L169)) is useful when upgrading to 3.9.0, but IMHO there also needs to be a way for it to be suppressed until libraries such as aiohttp-security can be updated, see e.g. [this](https://github.com/aio-libs/aiohttp-security/blob/c066a3a6/aiohttp_security/api.py#L127). I guess this could be done either by making a `UserWarning` subclass for which one could then make a filter, or by providing a configuration option.

**Comment by Dreamsorcerer** [comment]
It's a warning, so normal warning filters can be used (CLI argument or `warnings.filterwarnings()`). I'll try it out shortly and see if we can add to the documentation.

**Comment by Dreamsorcerer** [comment]
Hmm, maybe I misunderstood the module parameter, it doesn't look like I can filter it with `-Wignore::UserWarning:aiohttp.web_app` for example. Maybe a subclass makes more sense then.

Also, the warning was not meant to be displayed by default (only when the user enables warnings with `-Wdefault` or `-We`, for example). So, I need to review that too.

**Comment by haukex** [comment]
Ah, I see. I run my test suite with `PYTHONWARNINGS=error` which is why I noticed it immediately. I was apparently able to suppress it with `warnings.filterwarnings('ignore', category=UserWarning, message='It is recommended to use web.AppKey instances for keys.')`, but I feel like that's not optimal; if it's just meant to a help in upgrading then it should probably be opt-in.

**Comment by Dreamsorcerer** [comment]
Hmm, well ignoring it via CLI won't be so easy: https://github.com/python/cpython/issues/66733
Doesn't look like there's any way to do it on CLI without message matching. Should be able to do it with runtime code though.

**Comment by Dreamsorcerer** [comment]
OK, subclassed and ignored by default in #7677.

Should be able to separately ignore it with `-W 'ignore:It is recommended to use web.AppKey'` or `warnings.filterwarnings("ignore", category=web.NotAppKeyWarning)`.

**Comment by haukex** [comment]
Thanks! Other than that no issues so far, my test suite passes on 3.9.0b0 and Python 3.12; I just switched over to using `web.AppKey`. Keeping a set of keys in an enum apparently requires a bit of extra code though due to https://github.com/python/typing/issues/535, but at least it works:

```python
class AppKeys(enum.Enum):
    FOO = web.AppKey('foo', Foo)
    BAR = web.AppKey('bar', Bar)
    @property
    def value(self) -> web.AppKey:
        return super().value  # type: ignore
...
app[AppKeys.FOO.value] = Foo()
# etc.
```

**Comment by haukex** [comment]
Another place that makes use of string app keys and causes the warning is https://github.com/aio-libs/aiohttp-devtools/blob/master/aiohttp_devtools/runserver/serve.py

**Comment by Dreamsorcerer** [comment]
> Keeping a set of keys in an enum apparently requires a bit of extra code though due to [python/typing#535](https://github.com/python/typing/issues/535), but at least it works

That still throws away the typing information, making it pointless to use AppKey().

`reveal_type(app[AppKeys.FOO.value])  # Revealed type is "Any"`

Compared with:
```
foo_key = web.AppKey("foo", Foo)
reveal_type(app[foo_key])  # Revealed type is "foo.Foo"
```

**Comment by benz0li** [comment]
My [JupyterLab Python docker stack](https://github.com/b-data/jupyterlab-python-docker-stack) seems to work fine.

Build chain:

1. `docker build --build-arg PYTHON_VERSION=3.12.0 -t python/ver:3.12.0 -f `[`ver/latest.Dockerfile`](https://github.com/b-data/python-docker-stack/blob/a76f36ddb853c430e1f876bb283a62b6fadddfca/ver/latest.Dockerfile)` .`
2. `cd base && docker build --build-arg BUILD_ON_IMAGE=python/ver --build-arg PYTHON_VERSION=3.12.0 -t jupyterlab/python/base:3.12.0 -f `[`latest.Dockerfile`](https://github.com/b-data/jupyterlab-python-docker-stack/blob/52ccdbf6750e6e68d72c2b30fb1f9d6447894697/base/latest.Dockerfile)` .`
    * `git diff base/latest.Dockerfile`
      ```diff
      diff --git a/base/latest.Dockerfile b/base/latest.Dockerfile
      index 1a8ff15..40effc9 100644
      --- a/base/latest.Dockerfile
      +++ b/base/latest.Dockerfile
      @@ -7,7 +7,7 @@ ARG CUDA_IMAGE_FLAVOR
       ARG NB_USER=jovyan
       ARG NB_UID=1000
       ARG JUPYTERHUB_VERSION=4.0.2
      -ARG JUPYTERLAB_VERSION=3.6.6
      +ARG JUPYTERLAB_VERSION=4.0.6
       ARG CODE_BUILTIN_EXTENSIONS_DIR=/opt/code-server/lib/vscode/extensions
       ARG CODE_SERVER_VERSION=4.17.1
       ARG GIT_VERSION=2.42.0
      @@ -227,12 +227,14 @@ RUN mkdir /opt/code-server \
       ## Install JupyterLab
       RUN export PIP_BREAK_SYSTEM_PACKAGES=1 \
         && pip install \
      +    aiohttp==3.9.0b0 \
           jupyter-server-proxy \
           jupyterhub==${JUPYTERHUB_VERSION} \
           jupyterlab==${JUPYTERLAB_VERSION} \
      -    jupyterlab-git \
      +    #jupyterlab-git \
           jupyterlab-lsp \
           notebook \
      +    nbclassic \
           nbconvert \
           python-lsp-server[all] \
         ## Include custom fonts
      ```

Run container:

```bash
docker run -it --rm -p 8888:8888 jupyterlab/python/base:3.12.0
```

Enjoy!

**Comment by Dreamsorcerer** [comment]
One more issue I've found while testing is with using websockets. If we have open websockets, then the graceful shutdown will always wait for the timeout (as the websocket tasks won't end) before continuing with shutdown.

Should we add a pre_shutdown signal or something? Then websocket closing (as shown in https://docs.aiohttp.org/en/stable/web_advanced.html#graceful-shutdown) can be done in a pre-shutdown handler before the server waits for running tasks to complete.

**Comment by haukex** [comment]
> That still throws away the typing information

Hm, mypy doesn't see it, though the type hinting in VSCode does:
![Unbenannt](https://github.com/aio-libs/aiohttp/assets/4613111/32fd66e3-4280-4a9e-a019-6fad09649316)

**Comment by Dreamsorcerer** [comment]
> Hm, mypy doesn't see it, though the type hinting in VSCode does:

Isn't that a hardcoded annotation? It should work without an annotation.
That issue you linked, and the workaround you implemented, only narrow the value types to AppKey. There's no way to get the precise values unless something is tracking the individual enum values.

**Comment by hyperlogic** [comment]
When trying to `pip install -U --pre aiohttp` on Windows 11 using Visual Studio 2022 and Python 3.12.  I get the following error:

```
      aiohttp/_websocket.c(196): fatal error C1083: Cannot open include file: 'longintrepr.h': No such file or directory
      error: command 'C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.37.32822\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2
      [end of output]
 ```

**Comment by Malogna** [comment]
> When trying to `pip install -U --pre aiohttp` on Windows 11 using Visual Studio 2022 and Python 3.12. I get the following error:
> 
> ```
>       aiohttp/_websocket.c(196): fatal error C1083: Cannot open include file: 'longintrepr.h': No such file or directory
>       error: command 'C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.37.32822\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2
>       [end of output]
> ```

See PyPI, [aiohttp 4.0.0a1](https://pypi.org/project/aiohttp/#history) is a pre-build from 2019. Use `pip install aiohttp==3.9.0b0` to get Python 3.12 working with aiohttp

**Comment by haukex** [comment]
It's not obvious from the screenshot but that annotation isn't hardcoded, it's auto generated by Pylance because I have its "Inlay Hints: Variable Types" featured turned on. Here's a better screenshot:

![Unbenannt](https://github.com/aio-libs/aiohttp/assets/4613111/a18bdfba-69b8-4b79-ae3e-7cc73e85c749)

In my actual app I have it split across multiple files so it really does pick up on it correctly.

But anyway, since mypy isn't picking up on it, I guess I'll have to switch away from an enum.

**Comment by hyperlogic** [comment]
Thanks! aiohttp 4.0.0a1 correctly installs for me

**Comment by nolar** [comment]
Not directly related to the functionality, but to packaging: In 3.9.0b0, the `tests` directory is packaged into the wheels and installed into `site-packages` as a package. I guess, this is not intended. Besides, it makes PyCharm go crazy and navigate my code improperly.

**Comment by theCapypara** [comment]
> See PyPI, [aiohttp 4.0.0a1](https://pypi.org/project/aiohttp/#history) is a pre-build from 2019. Use `pip install aiohttp==3.9.0b0` to get Python 3.12 working with aiohttp

As a suggestion, it would probably be a good idea to yank the 4.0 pre-releases if they are 4 years old and no longer relevant to what 4.0 would look like to avoid confusion like this

**Comment by nolar** [comment]
As for the functionality, I have tried the beta in my project — Kopf — where it is used  and covered by tests extensively:

* As an aiohttp client talking to a local Kubernetes server, with simple & streaming requests.
* As an aiohttp server for webhooks (the app listens, the tests or K8s connect to it).
* In tests, via `aresponses`, which is based on `aiohttp.BaseTestServer` to run a local web server and intercept all outgoing web requests (even to 3rd party sites) into the mock server. The clients (tests) are also aiohttp, of course.

All works well: https://github.com/nolar/kopf/pull/1067

I had to fix one thing in https://github.com/nolar/kopf/pull/1069 — but that is a case of "one man's bug is another man's feature": I relied on server not waiting for connections to be closed. When they started taking 15s (keep-alive timeout) to close, each test started to take 15s and that was a problem. I changed the order of disconnecting: the clients terminate first and close their keep-alive connections, so the server is now able to exit instantly.

**Comment by Pliner** [comment]
>I had to fix one thing in https://github.com/nolar/kopf/pull/1069 — but that is a case of "one man's bug is another man's feature": I relied on server not waiting for connections to be closed. When they started taking 15s (keep-alive timeout) to close, each test started to take 15s and that was a problem. I changed the order of disconnecting: the clients terminate first and close their keep-alive connections, so the server is now able to exit instantly.

I seem to have the same issue with tests as @nolar mentioned:

**aiohttp 3.8.6**
```
439 passed in 24.67s
```

**aiohttp 3.9.0b0**
```
439 passed in 352.14s
```

Commenting [this with block with its content](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/web_runner.py#L86) makes tests duration the same for both versions.

I haven't found yet what stops aiohttp to shutdown in my case.

Meanwhile, @Dreamsorcerer What do you think about allowing to set `shutdown_timeout` [here](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/test_utils.py#L128)? It will help to override its default value in tests.

`aiohttp_client` (pytest fixture from https://github.com/aio-libs/pytest-aiohttp) supports passing `server_kwargs` to `TestServer` (a subclass of `BaseTestServer`), so `shutdown_timeout` might be added to `BaseTestServer` constructor and used to construct `SockSite`.

**Comment by nolar** [comment]
@Pliner In my case those were keepalive connections.

Setting force_close=True in the client side (TCPConnector) solved the problem. But I do want keepalives normally.

Manipulating the keepalive timeout clearly reflected to the test duration. (Default: 15s.)

In tests, I reordered pytest fixtures so that the client teardown happens before the server teardown (by making the client fixture dependent on the server fixture). Luckily, I control both sides here. 

But indeed, it would be helpful to ask the server force-terminate itself without waiting for connections to be closed from the other side.

**Comment by Dreamsorcerer** [comment]
> But indeed, it would be helpful to ask the server force-terminate itself without waiting for connections to be closed from the other side.

Well, that's the shutdown_timeout, right? Just reduce it to something like 0 or 1 if you don't care about handlers getting cancelled.

Although, my understanding is that the server is waiting on pending tasks, not open connections. So, I don't think a keep-alive connection should make a difference. Maybe I need to take a closer look at what happens with your code, but is it not the case that it is actively waiting on a response from an endpoint or websocket?

If it's a case that we want to tell those connections to shutdown gracefully, then that's the issue I highlighted already: https://github.com/aio-libs/aiohttp/issues/7675#issuecomment-1752143748
My suggestion is that we add a pre_shutdown signal.

**Comment by Dreamsorcerer** [comment]
> Meanwhile, @Dreamsorcerer What do you think about allowing to set `shutdown_timeout` [here](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/test_utils.py#L128)? It will help to override its default value in tests.

As a configurable value, or simply set it 0 for tests (as we have handler_cancellation set to True)?
I think it's probably something that's not worth having in tests (your test code should be finished with everything when you want to cleanup the server, so why would we want to wait on anything still hanging around), so I'm thinking just setting it to 0 is probably a good choice.

**Comment by nolar** [comment]
I will try the shutdown timeout and will make an isolated repro this weekend.

**Comment by Pliner** [comment]
@nolar @Dreamsorcerer 

It looks like I managed to reproduce the issue on my side. There is a simplified example which hangs for 15 secs on teardown with aiohttp 3.9.0b0:

```
from typing import AsyncIterator

import aiohttp.web
import pytest


@pytest.fixture
async def app(event_loop):
    async def handler(request: aiohttp.web.Request) -> aiohttp.web.Response:
        return aiohttp.web.json_response(data=42)

    async def setup_client_session(app: aiohttp.web.Application) -> AsyncIterator:
        app["client_session"] = aiohttp.ClientSession()
        yield
        await app["client_session"].close()

    app = aiohttp.web.Application()
    app.cleanup_ctx.append(setup_client_session)
    app.router.add_get("/answer", handler)
    yield app


@pytest.fixture
async def simple_server_client(aiohttp_client, app):
    return await aiohttp_client(app)


class SimpleServerClient:
    def __init__(self, client_session: aiohttp.ClientSession, base_url: str):
        self._client_session = client_session
        self._base_url = base_url

    async def get_answer(self) -> int:
        async with self._client_session.get(f"{self._base_url}/answer") as response:
            return await response.json()


@pytest.fixture
async def typed_simple_server_client(simple_server_client):
    yield SimpleServerClient(
        simple_server_client.server.app["client_session"],  # a server aiohttp.ClientSession is used
        f"http://{simple_server_client.server.host}:{simple_server_client.server.port}",
    )


async def test_reproduce_slow_shutdown(typed_simple_server_client, simple_server_client):
    assert await typed_simple_server_client.get_answer() == 42
```

Here is a pytest log:

```
Testing started at 11:51 ...
Launching pytest with arguments tests/test_reproduce_slow_shutdown.py --no-header --no-summary -q in tests

============================= test session starts ==============================
collecting ... collected 1 item

test_reproduce_slow_shutdown.py::test_reproduce_slow_shutdown DEBUG:asyncio:Using selector: KqueueSelector
DEBUG:asyncio:Using selector: KqueueSelector
PASSED     [100%]INFO:aiohttp.access:127.0.0.1 [12/Oct/2023:10:51:24 +0000] "GET /answer HTTP/1.1" 200 161 "-" "Python/3.11 aiohttp/3.9.0b0"
DEBUG:asyncio:Using selector: KqueueSelector


======================== 1 passed, 1 warning in 15.72s =========================
```

Could you confirm that it hangs on your side?


Changing `typed_simple_server_client` fixture, so that it creates a new ClientSessions and doesn't use a server one, resolves the issue:

```
@pytest.fixture
async def typed_simple_server_client(simple_server_client):
    async with aiohttp.ClientSession() as client_session:
        yield SimpleServerClient(
            client_session,
            f"http://{simple_server_client.server.host}:{simple_server_client.server.port}",
        )
```

I don't know why it was decided to reuse server ClientSession in tests, however I don't have a good explanation why it hangs.

**Comment by Dreamsorcerer** [comment]
OK, there is a task running for RequestHandler.start(), which is waiting at https://github.com/aio-libs/aiohttp/blob/master/aiohttp/web_protocol.py#L515
So, it appears to wait for incoming data. This means that a keep-alive connection, if actively making more requests, could pause the server shutdown until the timeout is reached. Maybe there's a way we can automatically close any idle connections...

**Comment by Pliner** [comment]
> OK, there is a task running for RequestHandler.start(), which is waiting at https://github.com/aio-libs/aiohttp/blob/master/aiohttp/web_protocol.py#L515 So, it appears to wait for incoming data. This means that a keep-alive connection, if actively making more requests, could pause the server shutdown until the timeout is reached. Maybe there's a way we can automatically close any idle connections...

@Dreamsorcerer I see, thanks. 

Please correct me if I am wrong, but it looks like the same issue might happen with any long lived connections created on-demand (I mean not on start up) during requests execution?

**Comment by Pliner** [comment]
A small example to the previous comment using aiormq, when a connection is created on demand. Please run rabbitmq locally via `docker run -p 15672:15672 -p 5672:5672 rabbitmq:management`.

```
import aiohttp.web
import aiormq
import pytest


class RmqPublisher:
    def __init__(self):
        self._connection = None
        self._channel = None

    async def publish(self):
        if self._connection is None:
            self._connection = await aiormq.connect("amqp://guest:guest@localhost/")
        if self._channel is None:
            self._channel = await self._connection.channel()
        await self._channel.basic_publish(bytes(42), routing_key="", exchange="")

    async def close(self):
        if self._channel is not None:
            await self._channel.close()
            self._channel = None
        if self._connection is not None:
            await self._connection.close()
            self._connection = None


@pytest.fixture
async def app(event_loop):
    async def handler(request):
        await request.app["rmq_publisher"].publish()
        return aiohttp.web.json_response()

    async def setup_client_session(app):
        app["rmq_publisher"] = RmqPublisher()
        yield
        await app["rmq_publisher"].close()

    app = aiohttp.web.Application()
    app.cleanup_ctx.append(setup_client_session)
    app.router.add_get("/publish", handler)
    yield app


@pytest.fixture
async def simple_server_client(aiohttp_client, app):
    return await aiohttp_client(app)


class SimpleServerClient:
    def __init__(self, client_session, base_url):
        self._client_session = client_session
        self._base_url = base_url

    async def publish(self) -> None:
        async with self._client_session.get(f"{self._base_url}/publish") as response:
            response.raise_for_status()


@pytest.fixture
async def typed_simple_server_client(simple_server_client):
    async with aiohttp.ClientSession() as client_session:
        yield SimpleServerClient(client_session, f"http://{simple_server_client.host}:{simple_server_client.port}")


async def test_reproduce_slow_shutdown(typed_simple_server_client, simple_server_client):
    await typed_simple_server_client.publish()
```

Here is a pytest log.

```
Testing started at 16:11 ...
Launching pytest with arguments /Users/iuriipliner/Work/aio-request/tests/test_reproduce_slow_shutdown.py --no-header --no-summary -q in /Users/iuriipliner/Work/aio-request/tests

============================= test session starts ==============================
collecting ... collected 1 item

test_reproduce_slow_shutdown.py::test_reproduce_slow_shutdown DEBUG:asyncio:Using selector: KqueueSelector
DEBUG:asyncio:Using selector: KqueueSelector
PASSED     [100%]DEBUG:aiormq.connection:Connecting to: amqp://guest:******@localhost:5672/
DEBUG:aiormq.connection:Prepare to send ChannelFrame(payload=b'\x01\x00\x01\x00\x00\x00\x06\x00\x14\x00\n\x010\xce', should_close=False, drain_future=None)
DEBUG:aiormq.connection:Received frame <Channel.OpenOk object at 0x10562f5d0> in channel #1 weight=16 on <Connection: "amqp://guest:******@localhost:5672/" at 0x105612710>
DEBUG:aiormq.connection:Prepare to send ChannelFrame(payload=b'\x01\x00\x01\x00\x00\x00\x05\x00U\x00\n\x00\xce', should_close=False, drain_future=None)
DEBUG:aiormq.connection:Received frame <Confirm.SelectOk object at 0x105685910> in channel #1 weight=12 on <Connection: "amqp://guest:******@localhost:5672/" at 0x105612710>
DEBUG:aiormq.connection:Prepare to send ChannelFrame(payload=b'\x01\x00\x01\x00\x00\x00\t\x00<\x00(\x00\x00\x00\x00\x00\xce\x02\x00\x01\x00\x00\x000\x00<\x00\x00\x00\x00\x00\x00\x00\x00\x00*\x10\x80\x01 c3cc7cc9ca5141e79f9444bf105d9fdf\xce\x03\x00\x01\x00\x00\x00*\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xce', should_close=False, drain_future=<Future pending cb=[FutureStore.__on_task_done.<locals>.remover() at /Users/iuriipliner/.pyenv/versions/aio-request/lib/python3.11/site-packages/aiormq/base.py:33, <1 more>, Task.task_wakeup()]>)
DEBUG:aiormq.connection:Received frame <Basic.Ack object at 0x105634c30> in channel #1 weight=21 on <Connection: "amqp://guest:******@localhost:5672/" at 0x105612710>
INFO:aiohttp.access:127.0.0.1 [12/Oct/2023:15:11:21 +0000] "GET /publish HTTP/1.1" 200 144 "-" "Python/3.11 aiohttp/3.9.0b0"
DEBUG:aiormq.connection:Prepare to send ChannelFrame(payload=b'\x08\x00\x00\x00\x00\x00\x00\xce', should_close=False, drain_future=None)
DEBUG:aiormq.connection:Received frame <pamqp.heartbeat.Heartbeat object at 0x105602110> in channel #0 weight=8 on <Connection: "amqp://guest:******@localhost:5672/" at 0x105612710>
DEBUG:aiormq.connection:Prepare to send ChannelFrame(payload=b'\x08\x00\x00\x00\x00\x00\x00\xce', should_close=False, drain_future=None)
DEBUG:aiormq.connection:Received frame <pamqp.heartbeat.Heartbeat object at 0x105615490> in channel #0 weight=8 on <Connection: "amqp://guest:******@localhost:5672/" at 0x105612710>
DEBUG:aiormq.connection:Prepare to send ChannelFrame(payload=b'\x01\x00\x01\x00\x00\x00\x0b\x00\x14\x00(\x00\xc8\x00\x00\x00\x00\x00\xce', should_close=False, drain_future=None)
DEBUG:aiormq.connection:Received frame <Channel.CloseOk object at 0x105615110> in channel #1 weight=12 on <Connection: "amqp://guest:******@localhost:5672/" at 0x105612710>
DEBUG:aiormq.connection:Closing connection <Connection: "amqp://guest:******@localhost:5672/" at 0x105612710> cause: <class 'asyncio.exceptions.CancelledError'>
DEBUG:aiormq.connection:Sending <Connection.Close object at 0x1055f5070> to <Connection: "amqp://guest:******@localhost:5672/" at 0x105612710>
DEBUG:aiormq.connection:Reader exited for <Connection: "amqp://guest:******@localhost:5672/" at 0x105612710>
DEBUG:aiormq.connection:Writer on connection amqp://guest:******@localhost:5672/ closed
DEBUG:aiormq.connection:Writer exited for <Connection: "amqp://guest:******@localhost:5672/" at 0x105612710>
DEBUG:aiormq.connection:Closing connection <Connection: "amqp://guest:******@localhost:5672/" at 0x105612710> cause: CancelledError()
DEBUG:asyncio:Using selector: KqueueSelector


=================== 1 passed, 1 warning in 60.05s (0:01:00) ====================
```

In this case shutdown took 60 secs (a shutdown_timeout, to be precise).

**Comment by frenck** [comment]
We ran into a regression with 3.9.0b0 (including 3.8.6), which thus blocks the Home Assistant project testing any further with this beta and from moving forward with Python 3.12.

Test that confirms the regression @ this end: #7697 
Which originates from `llhttp`, test to show it on their end: https://github.com/nodejs/llhttp/pull/256

Refs:
- https://github.com/home-assistant/core/issues/101893  (issue)
- https://github.com/home-assistant/core/pull/101913 (downgrade to 3.8.5 for our stable release)

**Comment by dlroy88osu** [comment]
Python 3.12 support. openai api failing due to dependency.
https://github.com/openai/openai-python/issues/654

**Comment by Dreamsorcerer** [comment]
> Python 3.12 support. openai api failing due to dependency. [openai/openai-python#654](https://github.com/openai/openai-python/issues/654)

I don't understand your comment. Is there an issue you are finding with the beta release?

**Comment by Dreamsorcerer** [comment]
> It looks like I managed to reproduce the issue on my side. There is a simplified example which hangs for 15 secs on teardown with aiohttp 3.9.0b0:

Thanks, I recreated the test in the same style as our existing shutdown tests in #7699.

**Comment by felixxm** [comment]
Django checked :heavy_check_mark:, Thanks :+1: (https://github.com/django/django/pull/17331.)

**Comment by nolar** [comment]
Here is an isolated repro:

It takes 15s. BUT: if the client is not aiohttp, e.g. it is Kubernetes itself doing a webhook request to the webhook based on aiohttp (as in typical e2e tests), so it does NOT close the connection from the client side in 15s (here: simulated in 33s), the server does not exit until then — regardless of `shutdown_timeout`. In my particular case, it leads to test timeouts even with rather generous timeout of 60s; but it can be even longer (hours?).

Re `shutdown_timeout` — it is used a few lines below in `BaseSite.stop()`, where it waits for some tasks. However, the delay happens at `await self._server.wait_closed()` and is not constrained by the shutdown timeout.

```python
        # named pipes do not have wait_closed property
        if hasattr(self._server, "wait_closed"):
            await self._server.wait_closed()  # ❓ unrestricted by time

        # Wait for pending tasks for a given time limit.
        with suppress(asyncio.TimeoutError):
            await asyncio.wait_for(
                self._wait(asyncio.current_task()), timeout=self._shutdown_timeout
            )
```

**Expected:** It should take near-zero time, since the `shutdown_timeout` is set to near-zero (I saw `delay is None or delay <= 0` in the code, so I think 0 means unlimited).

**The repro:**

```python
# aiohttp==3.9.0b0
import aiohttp.web, asyncio


async def _serve_fn(request: aiohttp.web.Request) -> aiohttp.web.Response:
    pass


async def main():
    app = aiohttp.web.Application()
    app.add_routes([aiohttp.web.post("/", _serve_fn)])
    runner = aiohttp.web.AppRunner(app, handle_signals=False)
    await runner.setup()
    try:
        site = aiohttp.web.TCPSite(runner, None, 12345, reuse_port=True,
                                   shutdown_timeout=0.0001)
        await site.start()

        # Initiate a keepalive connection.
        session = aiohttp.ClientSession(
            connector=aiohttp.TCPConnector(
                limit=0,
                # force_close=True,  # ✅ this helps
                keepalive_timeout=33,  #  👿simulate an evil client, e.g. K8s itself
            ),
        )
        resp = await session.get(f'http://127.0.0.1:{site._port}/')
        text = await resp.read()
        print(f"{text=}")
        # await session.close()  # ✅ this helps

    finally:
        await runner.cleanup()  # ❌ this fails

    # Keep the session LONGER than the server.
    await session.close()

asyncio.run(main())
```


Run it:

```
$ time python _serverstop.py 
text=b'405: Method Not Allowed'

real	0m34.112s
user	0m0.277s
sys	0m0.082s
```

**Comment by Dreamsorcerer** [comment]
See comment above, should be fixed in #7699. I'll also reduce shutdown_timeout in tests in another PR.

> However, the delay happens at `await self._server.wait_closed()` and is not constrained by the shutdown timeout.

I think something must be wrong in your testing though, I don't think there's any chance that it was blocked there. All that does is ensure that the server is not listening anymore, it has nothing to do with open connections:
https://docs.python.org/3/library/asyncio-eventloop.html?highlight=wait_closed#asyncio.Server.close

> (I saw `delay is None or delay <= 0` in the code, so I think 0 means unlimited).

Not sure where you saw that, the documentation states `None` is unlimited. When I tried before with `0`, it times out immediately.
https://docs.python.org/3/library/asyncio-task.html#asyncio.wait_for

**Comment by nolar** [comment]
>  I don't think there's any chance that it was blocked there

I have made this conclusion from numerous print() statements that I put in aiohttp’s code on every line and in every if. 😅 The repro is hanging on wait_closed() for sure. You can try it.

**Comment by nolar** [comment]
> Not sure where you saw that, the documentation states `None` is unlimited. When I tried before with `0`, it times out immediately.

That was a few lines below, in this call chain — so I was a bit cautious about strict `0`:

```
        await self._runner.server.shutdown(self._shutdown_timeout)
```

```
        coros = [conn.shutdown(timeout) for conn in self._connections]
```

```
            async with ceil_timeout(timeout):
```

```
    if delay is None or delay <= 0:
        return async_timeout.timeout(None)
```

**Comment by nolar** [comment]
> See comment above, should be fixed in https://github.com/aio-libs/aiohttp/pull/7699. I'll also reduce shutdown_timeout in tests in another PR.

@Dreamsorcerer Sorry, I have lost it in all the references to this issue from unrelated libraries.

Yes, indeed, it works!

I was not able to install that PR locally because of some C-level compilation errors on MacOS. But I have copied all the `.py` files into `site-packages`. With that state of the code, the repo is fast indeed. Uninstalling and re-installing aiohttp to 3.9.0b0 brings the issue back. So the issue can be considered solved (when released). Thanks!

**Comment by 2xB** [comment]
@Dreamsorcerer Just to mention that tomorrow, Fedora 39 will likely be released. This means that beginning from tomorrow, likely people will have issues due to aiohttp even without using the beta channel of a distro.

I do not know about the maintenance workload distribution of this project, it's of course great to see that this project takes beta testing seriously and in any case hope that this doesn't cause too much inconvenience!

**Comment by Dreamsorcerer** [comment]
> > I don't think there's any chance that it was blocked there
> 
> I have made this conclusion from numerous print() statements that I put in aiohttp’s code on every line and in every if. 😅 The repro is hanging on wait_closed() for sure. You can try it.

Something is unusual with your setup. I had to increase shutdown_timeout to reproduce the hang at all, and it never happened on that line... Well, atleast the issue is still resolved with the same fix.

**Comment by emontnemery** [comment]
There's a problem in how the `BaseRunner` and `BaseSite` determine which tasks to wait for when shutting down.
The implementation uses `asyncio.all_tasks()` in `BaseRunner.setup` and stores that in a `set`. `BaseSite._wait` then again calls `asyncio.all_tasks()` and waits for all tasks not in the original set.

This implementation is not correct:
- Storing references to tasks in a `set` prevents those tasks from being garbage collected
- Waiting for all tasks not in the original set will wait also wait for tasks not relevant to the `BaseRunner` being shut down. Tracking of relevant tasks for a graceful shutdown needs to be done manually, for example by:

    ```py
    class BaseRunner(ABC):

        tasks: set[asyncio.Task[Any]] = set()

        def add_task(self, task: asyncio.Task[Any]):
            self.tasks.add(task)
            self.task.add_done_callback(tasks.remove)

        def create_task(self, target: Coroutine[Any, Any, _R], name: str) -> asyncio.Task[_R]:
            task = self.loop.create_task(target, name=name)
            self.add_task(task)
            return task       
    ```

  With this change, tasks relevant for shutting down the runner gracefully should be created by calling `BaseRunner.create_task`, or if that's not possible, be added to the set of tasks by calling `BaseRunner.add_task`

Should I create a separate issue to track this?

**Comment by bdraco** [comment]
https://github.com/aio-libs/aiohttp/pull/7188/files#diff-01635c3f6f43968db496de09ad1eea883748ae07b7427acd7c32a7b75240287aR310

This is holding a strong reference to unrelated tasks. That's why we never see tasks garbage collected that were started before the runner anymore.

It also prevents the warning about `exception was never retrieved` from working: https://github.com/python/cpython/blob/bcc941bd4a7fbed3b20f5a5fc68b183fda1506a5/Lib/asyncio/futures.py#L99

**Comment by Dreamsorcerer** [comment]
> * Storing references to tasks in a `set` prevents those tasks from being garbage collected

A WeakSet should solve that.

> * Waiting for all tasks not in the original set will wait also wait for tasks not relevant to the `BaseRunner` being shut down. Tracking of relevant tasks for a graceful shutdown needs to be done manually, for example by:

This was intended to ensure that all spawned tasks have a chance to complete normally. For example, when handler cancellation is used, we recommend spawning new tasks into the background to avoid cancellation. But, if we don't allow for the waiting, then we end up cancelling those background tasks at shutdown anyway. As the point of making them background tasks was to ensure they would complete (e.g. making sure some DB updates go through), the idea was to give them a grace period on shutdown to actually complete.

Do you have an example of how this is causing an issue?

**Comment by bdraco** [comment]
A WeakSet should solve what I posted above 👍

**Comment by emontnemery** [comment]
> This was intended to ensure that all spawned tasks have a chance to complete normally. For example, when handler cancellation is used, we recommend spawning new tasks into the background to avoid cancellation. But, if we don't allow for the waiting, then we end up cancelling those background tasks at shutdown anyway. As the point of making them background tasks was to ensure they would complete (e.g. making sure some DB updates go through), the idea was to give them a grace period on shutdown to actually complete.

Sure, the intention of waiting for tasks to finish is clear. However, the method used waits for every single task in the event loop to finish, not for tasks which are related to the aiohttp instance being shut down.

> Do you have an example of how this is causing an issue

The shutdown method will deadlock until the timeout expires if what's being shut down is itself being tracked by another task, for example by `asyncio.gather` which creates an outer task to track the result of all tasks being gathered: https://github.com/python/cpython/blob/v3.11.5/Lib/asyncio/tasks.py#L718-L838. Maybe this explains some cases where people are reporting slow shutdowns with the 3.9 beta?

**Comment by bdraco** [comment]
> * Waiting for all tasks not in the original set will wait also wait for tasks not relevant to the `BaseRunner` being shut down. Tracking of relevant tasks for a graceful shutdown needs to be done manually, for example by:
>   ```python
>   class BaseRunner(ABC):
>   
>       tasks: set[asyncio.Task[Any]] = set()
>   
>       def add_task(self, task: asyncio.Task[Any]):
>           self.tasks.add(task)
>           self.task.add_done_callback(tasks.remove)
>   
>       def create_task(self, target: Coroutine[Any, Any, _R], name: str) -> asyncio.Task[_R]:
>           task = self.loop.create_task(target, name=name)
>           self.add_task(task)
>           return task       
>   ```

>     
>   With this change, tasks relevant for shutting down the runner gracefully should be created by calling `BaseRunner.create_task`, or if that's not possible, be added to the set of tasks by calling `BaseRunner.add_task`

I think this is a good solution. Its safer and in line with what is recommended on the `asyncio.Task.create_task` docs: https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task

**Comment by Dreamsorcerer** [comment]
> I think this is a good solution.

I just don't see how that can help with the original goal. Again, if a user spawns a task from a handler to e.g. ensure that something is written to DB, I want to ensure it has a chance to complete. I don't see any way the runner will know about that task other than waiting on all tasks.

> The shutdown method will deadlock until the timeout expires if what's being shut down is itself being tracked by another task, for example by `asyncio.gather`

So, you are using the runner directly, right? Do you think this feature is useful to you at all? I'm thinking that maybe I'll pull it back into run_app() and not include it in BaseRunner at all (I think I'll just need to add a callback parameter or something to the runner that run_app() can hook into).

As run_app() controls the event loop, I think the feature makes a lot more sense here than it perhaps does if you're using the runner directly in a bigger application.

**Comment by emontnemery** [comment]
> > I think this is a good solution.
> 
> I just don't see how that can help with the original goal. Again, if a user spawns a task from a handler to e.g. ensure that something is written to DB, I want to ensure it has a chance to complete. I don't see any way the runner will know about that task other than waiting on all tasks.

Right, that's precisely the problem; there's no way to automatically know if a task needs to be awaited or not, it's up to the user to make it work, for example with an API like I suggested previously. In Python 3.12 there are [task groups](https://docs.python.org/3/library/asyncio-task.html#task-groups), which are meant to solve exactly this kind of problem.

> > The shutdown method will deadlock until the timeout expires if what's being shut down is itself being tracked by another task, for example by `asyncio.gather`
> 
> So, you are using the runner directly, right? Do you think this feature is useful to you at all? I'm thinking that maybe I'll pull it back into run_app() and not include it in BaseRunner at all (I think I'll just need to add a callback parameter or something to the runner that run_app() can hook into).
> 
> As run_app() controls the event loop, I think the feature makes a lot more sense here than it perhaps does if you're using the runner directly in a bigger application.

Right, that sounds reasonable and I think it would solve the problem for Home Assistant. WDYT @bdraco ?

**Comment by bdraco** [comment]
> > So, you are using the runner directly, right? Do you think this feature is useful to you at all? I'm thinking that maybe I'll pull it back into run_app() and not include it in BaseRunner at all (I think I'll just need to add a callback parameter or something to the runner that run_app() can hook into).
> > As run_app() controls the event loop, I think the feature makes a lot more sense here than it perhaps does if you're using the runner directly in a bigger application.
> 
> Right, that sounds reasonable and I think it would solve the problem for Home Assistant. WDYT @bdraco ?

That sounds good. From a Home Assistant perspective we already have all are own task management, so we run our own shutdown methods anyways.  From that perspective, its best to have a way to disable aiohttp from trying to be helpful and waiting for the tasks since we currently would have to monkey patch out the `_wait` anyways.

Thanks!

**Comment by benz0li** [comment]
@Dreamsorcerer Is there an approximate timeframe for the [3.9](https://github.com/aio-libs/aiohttp/milestone/25) stable  release?

**Comment by Dreamsorcerer** [comment]
I've just added the last changes discussed above. Hoping another maintainer will review them shortly, otherwise I'll just have to push them through.

Once they're merged (all tracked in the top post), I'll push another beta release, but I think we can consider it stable for client-side users. Then I'll update the main aiohttp-* libraries to use AppKey, and if nobody has found any more issues by then we can promote it to a stable release.

**Comment by frenck** [comment]
This regression (shown by test) is still there though: <https://github.com/aio-libs/aiohttp/pull/7697>

Is there a way to reach out to the llhttp / nodejs team though? As the fix is pending here: <https://github.com/nodejs/llhttp/pull/256>

**Comment by Dreamsorcerer** [comment]
Ah, yes, ideally we want that included too. Though it's already in 3.8.6, so it's not a direct blocker.
I'd just ping them every week or 2, they seem to get back to these issues fairly quickly normally.

**Comment by frenck** [comment]
> Though it's already in 3.8.6, so it's not a direct blocker.

Well, 3.8.6 is broken too 🤷
It maybe should not have been in there in the first place.

> I'd just ping them every week or 2, they seem to get back to these issues fairly quickly normally.

Awesome 👍 

../Frenck

**Comment by Dreamsorcerer** [comment]
> > I don't think there's any chance that it was blocked there
> 
> I have made this conclusion from numerous print() statements that I put in aiohttp’s code on every line and in every if. 😅 The repro is hanging on wait_closed() for sure. You can try it.

I've ended up reproducing this. It seems that wait_closed() has, in fact, been doing nothing at all for all these years. In Python 3.12 it actually waits for all connections to complete. Given that aiohttp calls this method _before_ cancelling any connections, it just ends up hanging. So, this code has probably been wrong since aiohttp was first created, but worked due to a bug in the asyncio code.

**Comment by nolar** [comment]
@Dreamsorcerer This seems related:

* The triggering change: https://github.com/python/cpython/commit/5d09d11aa0b89aeba187f4f520728ccaf4fc5ac1 
* The discussion: https://github.com/python/cpython/issues/104344 

It seems, you are right, it was broken (no-op) in 3.7-3.11 but fixed in 3.12.

**Comment by sumitsahoo** [comment]
Nice progress everyone, fingers crossed for 3.9 release 🤞🏻

**Comment by CyberFlameGO** [comment]
What are the current blockers preventing 3.9 from being properly released?

**Comment by bdraco** [comment]
> What are the current blockers preventing 3.9 from being properly released?

- Shutdown fixes discussed above
- llhttp regression https://github.com/nodejs/llhttp/pull/256

**Comment by bdraco** [comment]
Unfortunately it turns out the breaking change in llhttp was intentional to fix rfc compliance. This means according to the rfc, the problem is in aiohttp and it's likely implemented that way because llhttp implemented it incorrectly originally.

https://github.com/nodejs/llhttp/pull/256#issuecomment-1782909852

For backwards compatibility, when using llhttp >= 9.x, I think we can remove the transfer-encoding header when it's a 304 (and the other codes specified in the rfc) for the client and server implementation.

This also means that 3.8.6 has the llhttp breaking change for server side 304s with transfer-encoding set, which makes sense why Home Assistant had to revert the upgrade from 3.8.5 to 3.8.6

**Comment by bdraco** [comment]
I’m going to look into what it will take to make those changes in aiohttp, verify if they will be relatively safe, get a test implementation together, and than validate it against the known failure case we have in Home Assistant. 

This is potential 2 changes:

- [ ] Server Side: Ignore body / chunked when `self.status in (204, 304) or 100 >= self.status < 200` https://datatracker.ietf.org/doc/html/rfc9112#section-6.3 - WIP https://github.com/aio-libs/aiohttp/pull/7756
- [ ] Client Side: Ignore body / chunked when `self.status in (204, 304) or 100 >= self.status < 200`  (Adjust pure python header parser to match llhttp handling of 204/304/1xx) https://datatracker.ietf.org/doc/html/rfc9112#section-6.3 - ready for review https://github.com/aio-libs/aiohttp/pull/7755

**Comment by bdraco** [comment]
The above PRs will fix the invalid chunked bodies aka `0\r\n\r\n` on 204/304/1xx and make the py parser behave the same as the c parser.

I'm going to do a lot more testing before I clean them up and mark them ready

**Comment by lvillis** [comment]
Eager for v3.9! Any ETA?

**Comment by bdraco** [comment]
Quick update on the linked PRs I'm working on above.

I'm mostly happy with the code now.

My current plan is

- [x] Today: More functional test on dev (backport to 3.8.6 squashed commit https://github.com/bdraco/aiohttp/tree/386_304_fixes) https://github.com/bdraco/aiohttp/commit/27ea3916b53d4b522512fe49c48d83112926e933. 
- [x] Monday: deploy to staging and functional testing on a staging environment  -- I've started on this a bit earlier than expected --
- [x] deploy mixed versions to staging to ensure backwards compat: I'll also need to test Home Assistant core running 3.8.5 and HA Supervisor running the back-ported patch
- [x] Monday: deploy to production and functional testing on a production environment 
- [x] Monday: Review logs and verify functional testing on production https://github.com/aio-libs/aiohttp/pull/7755
- [x] Tuesday: Review logs and verify functional testing on production rebase https://github.com/aio-libs/aiohttp/pull/7756 once the above PR is merged

If any of that goes better than expected it may be ready sooner, but usually things take more time than expected not less

**Comment by bdraco** [comment]
Testing is going well.  Everything looks good on staging now.  The 502 problems are gone as soon as the source server has the two PRs applied on top of 3.8.6

**Comment by bdraco** [comment]
Everything looks good in staging.  I'll put the changes on my production systems tomorrow.

**Comment by bdraco** [comment]
Deploying to production went well.  I'll review the logs in next 24 hours and make sure everything is ok, than I'll update the timeline/changelogs and mark the PRs ready for review if everything still looks OK

**Comment by chenrui333** [comment]
> Everything looks good in staging. I'll put the changes on my production systems tomorrow.

Thank you!! 💯

**Comment by Tinche** [comment]
Howdy,

I've been updating [some of my open source stuff](https://uapi.threeofwands.com/en/latest/) to 3.12, and as part of that I've upgraded to the 3.9 beta. Right away I noticed one of my test suites taking much longer than before.

I tracked the issue down to https://github.com/aio-libs/aiohttp/blob/3.9/aiohttp/web_runner.py#L90.

As a general rule of thumb, in an asyncio app you shouldn't assume all tasks are owned by aiohttp. I see there are provisions for this in the form of `runner.starting_tasks`, but it's not enough. I understand you want to give child tasks room to complete before cleaning up but this approach is too naive. A more robust approach would be just waiting on the tasks explicitly spawned for aiohttp, and teaching the users to use TaskGroups (easy nowdays).

I can work around my issue by playing around with the `shutdown_timeout` so it's not an urgent issue, but there's an opportunity to improve the architecture I wanted to bring up.

**Comment by bdraco** [comment]
@Tinche please see https://github.com/aio-libs/aiohttp/pull/7718

**Comment by bdraco** [comment]
I've pushed the latest 3.9 to production on a lot of different systems, and testing is going well (sans some unrelated regressions).  I'm going to update a few more systems in the morning.

**Comment by bdraco** [comment]
@Tinche Can you give the latest `3.9` branch a shot and see if its still an issue (https://aiohttp--7718.org.readthedocs.build/en/7718/web_advanced.html#graceful-shutdown)

**Comment by bdraco** [comment]
Testing went well overnight. Nothing showed up in the logs. I pushed 3.9 to everything else they I can test on

I can get this tested more widely by Home Assistant members if there is a 3.9.0b1 published

**Comment by Dreamsorcerer** [comment]
3.9.0b1 has been released, so please retest with that. Top post has been updated with current details.
(If last step of CI fails or needs approval, ping webknjaz: https://github.com/aio-libs/aiohttp/actions/runs/6747082136/job/18343149775)

I'll be away over the weekend, but next week I'll look at updating the aiohttp-* libraries to use AppKey, and if no more issues are reported, will probably aim to make a stable release the weekend after.

**Comment by bdraco** [comment]
Looks like the deploy needs an approval.  @webknjaz 
https://github.com/aio-libs/aiohttp/actions/runs/6747082136/job/18345995935

**Comment by Tinche** [comment]
I installed b1 from git, but it didn't address my issue.

**Comment by Dreamsorcerer** [comment]
> I installed b1 from git, but it didn't address my issue.

I've just looked through your code, and realised you're using the private _run_app() function. Given this is a private, undocumented function, I'm not really sure there's any expectation for this to work. If you don't want to use the main run_app() as your application entrypoint, then you should be using the app runners instead: https://docs.aiohttp.org/en/stable/web_advanced.html#application-runners

While you can probably tweak _run_app() to still work well for you (or we could make minor changes on our side), I would _expect_ us to break your code again sooner or later (something we've already done for people misusing these functions in a previous release).

**Comment by Dreamsorcerer** [comment]
> Looks like the deploy needs an approval. @webknjaz https://github.com/aio-libs/aiohttp/actions/runs/6747082136/job/18345995935

Hmm, must have missed that. Deployed now.

**Comment by bdraco** [comment]
Thanks. I'll get on to the wider testing today.

**Comment by Dreamsorcerer** [comment]
I've added a list of libraries to be updated before release (top post). If anybody wants to volunteer to update any of them, feel free to pick one from the list.

**Comment by Tinche** [comment]
@Dreamsorcerer Ooh, thanks for the tip, I'll switch to runners. I just need an async function to start running a server inside a broader asyncio app which might be running other things too.

**Comment by Dreamsorcerer** [comment]
Typically, we recommend making aiohttp the main app, run via run_app(), and then adding the other components of the app through cleanup_ctx: https://docs.aiohttp.org/en/stable/web_advanced.html#complex-applications
But, if that doesn't suit you, then yes, use the app runners (but, it'll be your job to manage the lifecycle of the app at that point).

**Comment by cbm755** [comment]
I dunno if you want +1's but my test suite passed on the new `b1` as well [1].  Thanks for your efforts on this.

[1] https://gitlab.com/plom/plom/-/merge_requests/2304

**Comment by bdraco** [comment]
I noticed regression which results in an increase in websocket latency while testing in production. I'm going to dig through the changes shortly and try to track down the cause

**Comment by bdraco** [comment]
It looks like the websocket is now compressing in the executor. For small messages the executor overhead seems to exceed the compression overhead.

This is was noticeable when using a web based terminal

**Comment by bdraco** [comment]
There is also still significant cpu run time even with compressing in the executor

**Comment by bdraco** [comment]
Oh. It's the flush. It spends more cpu time on the flush code than the actual compress code

**Comment by bdraco** [comment]
So there are two issues here


1. The flush isn't happening in the executor which means we still have significant cpu usage in the event loop

2. The executor overhead is higher than the compress and flush overhead for small packets which increases latency


To resolve this I propose:

- A new method that does the compress and flush in the executor in a single job. This should accomplish what I think the original goal was to move expensive run time out of the event loop. The single job should prevent a further increase in latency

- Not compressing small payloads in the executor where is likely the executor overhead exceeds the compress overhead. (Maybe payloads < 16k?)

**Comment by bdraco** [comment]
That seems to solve the issue in testing. I’ll prepare a PR later today

**Comment by bdraco** [comment]
https://github.com/aio-libs/aiohttp/pull/7223 Introduced the regression

**Comment by bdraco** [comment]
It looks like the goal was to avoid expensive run time for large payloads per https://github.com/aio-libs/aiohttp/issues/7192 and the websocket changes are collateral. Maybe it’s best to revert the change there for now, although I do see the value in running the compress in the executor for large payloads.

**Comment by bdraco** [comment]
I don't see a latency increase for http get requests with compression and small payloads though.

**Comment by bdraco** [comment]
Never mind. It's because Home Assistant already has a guard to prevent compressing small payloads.  As soon as I make the payload big enough I see it there as well. I think that's ok as the trade off seems worth it and it's easy to manage since an external caller can avoid compression. The websocket interface doesn't have that and latency is a lot more important to reduce there since it's used for things like terminals and sockets over websocket

**Comment by bdraco** [comment]
I was really hoping b1 was going to be the final one but I think we have to fix this before release as it's a 3 order of magnitude increase in executor job when testing Home Assistant on production.

**Comment by bdraco** [comment]
The smallest change to solve the issue is to simply increase `max_sync_chunk_size` in the websocket to 5KiB. I don't think we need to worry about the flush issue as its not a regression

https://github.com/aio-libs/aiohttp/pull/7797

**Comment by bdraco** [comment]
Found another issue in testing, it looks like there is an infinite loop in release connection
`     66 root      20   0 2117532 609476  77336 R  90.0   3.8  10:54.90 python3    `

<img width="1469" alt="Screenshot 2023-11-07 at 7 53 35 AM" src="https://github.com/aio-libs/aiohttp/assets/663432/c5ef6fd2-9eee-4ded-8095-261a92fb62aa">

**Comment by bdraco** [comment]
It looks like its coming in from https://github.com/aio-libs/aiohttp/pull/7764

**Comment by bdraco** [comment]
I _think_ I've fixed the loop in https://github.com/aio-libs/aiohttp/pull/7798

**Comment by binbjz** [comment]
> Found another issue in testing, it looks like there is an infinite loop in release connection ` 66 root 20 0 2117532 609476 77336 R 90.0 3.8 10:54.90 python3 `
> 
> <img alt="Screenshot 2023-11-07 at 7 53 35 AM" width="1469" src="https://user-images.githubusercontent.com/663432/281062706-c5ef6fd2-9eee-4ded-8095-261a92fb62aa.png">

@bdraco Hi, what is the name of this tool you are using？

**Comment by bdraco** [comment]
Its a cProfile that was converted to a callgrind with `pyprof2calltree` and opened in qcachegrind

https://www.home-assistant.io/integrations/profiler/#service-profilerstart

**Comment by alexrudd2** [comment]
Thank you for all your work getting 3.9 released!

Sad to report that https://github.com/aio-libs/aiohttp/pull/7780 breaks the test suite for `pymodbus`, which worked with `3.9.0b0`.  HomeAssistant is downstream of `pymodbus`, so I'm not sure if the HA problems are related.  @janiversen

https://github.com/pymodbus-dev/pymodbus/issues/1883

**Comment by binbjz** [comment]
@bdraco  Thank you very much.

**Comment by cooperlees** [comment]
Thanks for all your work!

We don't have huge tests for aiohttp, but black[d] seems happy with 3.9.0b1 - https://github.com/psf/black/pull/4035

**Comment by alexrudd2** [comment]
> #7780 breaks the test suite for `pymodbus`

Fixed by adjusting our test suite, with details [here](https://github.com/pymodbus-dev/pymodbus/issues/1883#issuecomment-1804196681).

@Dreamsorcerer 
The docs you linked are clear, but should this have been documented in `CHANGES`?  I don't see 7780 in there.

**Comment by Dreamsorcerer** [comment]
It's #7718, listed under 2 different points. I can try and add some warning or something in there. It might also be possible to make it more backwards-compatible by tracking the handlers for each site and having the site shutdown only those that belong to the site. But, I won't have time to do that before release.

**Comment by JCHacking** [comment]
It looks like [aioresponses](https://github.com/pnuckowski/aioresponses) is broken with 3.9.0b1

Example code:
```python
import asyncio
import json

import aiohttp
from aioresponses import aioresponses


async def main():
    # Create Session
    async with aiohttp.ClientSession() as session:
        # Mock responses
        with aioresponses() as m:
            m.add(
                method="GET",
                url="http://python.org",
                status=418,
                headers={},
                body=json.dumps({"content": "value"})
            )

            # Make request
            async with session.get('http://python.org') as response:
                print("Status:", response.status)
                print("Content-type:", response.headers['content-type'])
                html = await response.text()
                print("Body:", html[:15], "...")


asyncio.run(main())
```

Output with aiohttp 3.9.0b0:
```
Status: 418
Content-type: application/json
Body: {"content": "va ...
```

Output with aiohttp 3.9.0b1:
```
Status: 418
Content-type: application/json
Traceback (most recent call last):
  File "C:\Users\jchacking\PycharmProjects\NGCSQALIBS\py_requests\dev.py", line 29, in <module>
    asyncio.run(main())
  File "C:\Users\jchacking\AppData\Local\Programs\Python\Python312\Lib\asyncio\runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "C:\Users\jchacking\AppData\Local\Programs\Python\Python312\Lib\asyncio\runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jchacking\AppData\Local\Programs\Python\Python312\Lib\asyncio\base_events.py", line 664, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "C:\Users\jchacking\PycharmProjects\NGCSQALIBS\py_requests\dev.py", line 25, in main
    html = await response.text()
           ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jchacking\PycharmProjects\NGCSQALIBS\py_requests\venv\Lib\site-packages\aiohttp\client_reqrep.py", line 1112, in text
    await self.read()
  File "C:\Users\jchacking\PycharmProjects\NGCSQALIBS\py_requests\venv\Lib\site-packages\aiohttp\client_reqrep.py", line 1083, in read
    await self._wait_released()  # Underlying connection released
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jchacking\PycharmProjects\NGCSQALIBS\py_requests\venv\Lib\site-packages\aiohttp\client_reqrep.py", line 1041, in _wait_released
    await self._writer
TypeError: object Mock can't be used in 'await' expression
```

aioresponses version -> 0.7.4
Maybe with aiohttp 3.9 they should change this library?

**Comment by Dreamsorcerer** [comment]
> Maybe with aiohttp 3.9 they should change this library?

Yes, they'll need to make sure `_writer` is set to an AsyncMock or something similar. The change should be backwards-compatible, as that task has always existed, we just didn't await on it at this point in the code before.

**Comment by JCHacking** [comment]
I am trying to see if I can solve this on my own, so I can do my bit.
I see that the way aioresponses works is then returning a ClientResponse, if I make the attribute of the writer constructor to be AsyncMock instead of Mock I get this error trace.

```
Status: 418
Content-type: application/json
Traceback (most recent call last):
  File "C:\Users\jcmencia\PycharmProjects\NGCSQALIBS\py_requests\dev.py", line 29, in <module>
    asyncio.run(main())
  File "C:\Users\jcmencia\AppData\Local\Programs\Python\Python312\Lib\asyncio\runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "C:\Users\jcmencia\AppData\Local\Programs\Python\Python312\Lib\asyncio\runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jcmencia\AppData\Local\Programs\Python\Python312\Lib\asyncio\base_events.py", line 664, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "C:\Users\jcmencia\PycharmProjects\NGCSQALIBS\py_requests\dev.py", line 25, in main
    html = await response.text()
           ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jcmencia\PycharmProjects\NGCSQALIBS\py_requests\venv\Lib\site-packages\aiohttp\client_reqrep.py", line 1112, in text
    await self.read()
  File "C:\Users\jcmencia\PycharmProjects\NGCSQALIBS\py_requests\venv\Lib\site-packages\aiohttp\client_reqrep.py", line 1083, in read
    await self._wait_released()  # Underlying connection released
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jcmencia\PycharmProjects\NGCSQALIBS\py_requests\venv\Lib\site-packages\aiohttp\client_reqrep.py", line 1041, in _wait_released
    await self._writer
TypeError: object AsyncMock can't be used in 'await' expression
```

But however I see that if I assign the writer as AsyncMock()() it works. Why is it necessary to call the AsyncMock instance?

**Comment by JCHacking** [comment]
What still needs to be done to officially release version 3.9.0? Is there anything I can do to help speed it up?

**Comment by Dreamsorcerer** [comment]
Yes, pick a library to update in the checklist, from the top post (and let us know, so we don't work on the same one).

**Comment by JCHacking** [comment]
Perfect, I will try to update aiojobs

**Comment by JCHacking** [comment]
I have already updated aiojobs, I have opened the following PR. With what you see, let me know if something needs to be changed.
https://github.com/aio-libs/aiojobs/pull/460

**Comment by JCHacking** [comment]
As far as I have been able to see with the failure to pass the CI, aiojobs does not directly install aiohttp for ci environment, if not for ci installs pytest-aiohttp so put aiohttp 3.8.6 which is the version that has this specified (With >=) may have to be added to the list of libraries to update?
https://github.com/aio-libs/aiojobs/pull/460

I understand that no because the code does not really need to be changed. It's more a question of dependencies and since it's there with a >= it should work once version 3.9.0 is released?

**Comment by JCHacking** [comment]
Hi, I have tried with aiohttp-session and the new version of aiohttp (3.9.0b1) to pass the tests, if they pass all correctly it would mean that it does not have required changes?

**Comment by Dreamsorcerer** [comment]
> I have tried with aiohttp-session and the new version of aiohttp (3.9.0b1) to pass the tests

Ah, yes, it's only storing information on the request object, not on the app. If you could make a small PR to update the demo though, that looks like the only thing that needs updating:
```
demo/login_required_example.py:32:        app["user"] = user
demo/login_required_example.py:40:    user = request.app["user"]
```

**Comment by JCHacking** [comment]
Ready, updated example 👍🏻 
https://github.com/aio-libs/aiohttp-session/pull/943

**Comment by JCHacking** [comment]
Open PR for aiohttp-security
https://github.com/aio-libs/aiohttp-security/pull/671

We still need to approve these PRs and aiohttp 3.9 could be released? Or is there anything else I can help with?

**Comment by bdraco** [comment]
Pulled latest 3.9 and pushed to production.  So far everything is running well.

**Comment by bdraco** [comment]
~~I pushed to a few more systems and found a trace in the log~~

Edit: it looks like I accidentally deployed `master` instead of `3.9` to that system. 🤦 

retesting now.....

```
Nov 13 20:21:17 homeassistant hassio_supervisor[546]:     async with self.sys_websession.post(
Nov 13 20:21:17 homeassistant hassio_supervisor[546]:   File "/usr/local/lib/python3.11/site-packages/aiohttp/client.py", line 1125, in __aenter__
Nov 13 20:21:17 homeassistant hassio_supervisor[546]:     self._resp = await self._coro
Nov 13 20:21:17 homeassistant hassio_supervisor[546]:                  ^^^^^^^^^^^^^^^^
Nov 13 20:21:17 homeassistant hassio_supervisor[546]:   File "/usr/local/lib/python3.11/site-packages/aiohttp/client.py", line 417, in _request
Nov 13 20:21:17 homeassistant hassio_supervisor[546]:     self._loop, real_timeout.total, ceil_threshold=real_timeout.ceil_threshold
Nov 13 20:21:17 homeassistant hassio_supervisor[546]:                 ^^^^^^^^^^^^^^^^^^
Nov 13 20:21:17 homeassistant hassio_supervisor[546]: AttributeError: 'int' object has no attribute 'total'
```

Edit: all looks good when I actually deployed `3.9`. The timeout issue is only on `master`/4.x so nothing to see here

**Comment by Dreamsorcerer** [comment]
> Edit: all looks good when I actually deployed `3.9`. The timeout issue is only on `master`/4.x so nothing to see here

We might have backwards-compatibility in 3.9, but that still indicates user error. Timeouts should be ClientTimeout, not int:
https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.ClientSession

You should probably have got a type error in mypy. But, if it was passed to request(), then it would have been missed because of #4749.

**Comment by bdraco** [comment]
I've beat this up as much as I can. From my perspective, 3.9.0 seems good to go.

**Comment by Dreamsorcerer** [comment]
I've pushed 3.9.0rc0, which contains some fixes and performance improvements.

One of the fixes required a few changes to some of our tests, I'm expecting this will only affect our internal tests, but if a couple of people can confirm their test suite is still passing that'd be helpful.

If no further issues are found, I'll promote it to a stable release over the weekend (when I have some time to make releases for the other libraries).

**Comment by Dreamsorcerer** [comment]
Hmm, it didn't like the version number, will have to push another.

**Comment by bdraco** [comment]
Perfect.  I'll get some wider testing on rc0 once the wheels finish cooking & approval

https://github.com/home-assistant/core/pull/103507 (that PR was stuck due to the issue above)

**Comment by salty-horse** [comment]
Please add classifiers for 3.11 and 3.12 to `setup.cfg`, as support for 3.12 is one of the reasons people are waiting for this release.

**Comment by Dreamsorcerer** [comment]
See the duplicate issues and PRs that are all closed. It was decided not to add them until the test suite is stable and passing. It still doesn't pass for 3.11 or 3.12 yet, which means support is still considered somewhat experimental.

bdraco started to try and fix them in #7773.

**Comment by JrooTJunior** [comment]
Any news about release date?

**Comment by Dreamsorcerer** [comment]
> Any news about release date?

https://github.com/aio-libs/aiohttp/issues/7675#issuecomment-1810330313
